rasa/utils/tensorflow/layers.py

```diff
@@ -198,24 +188,149 @@ class DenseWithSparseWeights(tf.keras.layers.Dense):
+
         self.kernel_mask = tf.Variable(
             initial_value=kernel_mask, trainable=False, name="kernel_mask"
         )
 
+    def _random_mask_with_no_empty_rows(
+        self, num_rows: int, num_cols: int, num_connected_per_output: int
+    ) -> tf.Tensor:
+        """Creates a random mask with given number of connections per output.
+
+        Args:
+            num_rows: Number of rows in the mask
+            num_cols: Number of columns in the mask
+            num_connected_per_output: Number of ones in each column
+
+        Returns:
+            A random mask matrix
+        """
+        num_disconnected_per_output = num_rows - num_connected_per_output
+
+        # We create each column separately
+        # as [1 1 1 ... 1 0 0 0 ... 0] and then shuffle it randomly
+        kernel_mask_cols = [
+            tf.random.shuffle(
+                tf.concat(
+                    [
+                        tf.ones(
+                            shape=num_connected_per_output, dtype=self.kernel.dtype
+                        ),
+                        tf.zeros(
+                            shape=num_disconnected_per_output, dtype=self.kernel.dtype
+                        ),
+                    ],
+                    axis=0,
+                )
+            )
+            for _ in range(num_cols)
+        ]
+        return tf.transpose(tf.stack(kernel_mask_cols, axis=0))
+
+    def _missing_connections(
+        self, num_rows: int, num_cols: int, kernel_mask: tf.Tensor
+    ) -> tf.Tensor:
+        """Creates a mask with random connections that are missing.
+
+        The given `kernel_mask` is guaranteed to have at least one input connected to
+        each output, but not vice versa. Some inputs may not be connected to any
+        output, i.e. there may be empty columns in the mask. This function returns
+        another mask (to be added to the original) that contains exactly one 1 in each
+        of the columns that is empty in `kernel_mask`. The row in which each 1 appears
+        is chosen randomly.
+
+        Args:
+            num_rows: Number of rows in the mask
+            num_cols: Number of columns in the mask
+            kernel_mask: The kernel mask matrix of size (num_rows, num_cols) that may
+                contain empty columns
+
+        Returns:
+            A mask with the missing connections.
+        """
+        empty_columns = tf.squeeze(
+            tf.where(tf.equal(tf.reduce_sum(kernel_mask, axis=1), 0)), -1
+        ).numpy()
+        random_rows = np.random.randint(num_cols, size=len(empty_columns))
+        indices = np.transpose([empty_columns, random_rows])
+        extra_entries = tf.sparse.to_dense(
+            tf.SparseTensor(
+                indices,
+                np.ones(len(empty_columns), dtype=np.float),
+                (num_rows, num_cols),
+            )
+        )
+        return tf.cast(extra_entries, dtype=kernel_mask.dtype)
```

Comment:
it seems to me that non-empty random mask got lost

Reply:
It was added by `kernel_mask += self._missing_connections(num_rows, num_cols, kernel_mask)`